lenskit-eval/src/main/java/org/lenskit/eval/traintest/recommend/TopNNDPMMetric.java

```diff
@@ -0,0 +1,155 @@
+import org.grouplens.lenskit.util.statistics.MeanAccumulator;
+import org.lenskit.api.ResultList;
+import org.lenskit.eval.traintest.AlgorithmInstance;
+import org.lenskit.eval.traintest.DataSet;
+import org.lenskit.eval.traintest.TestUser;
+import org.lenskit.eval.traintest.metrics.MetricResult;
+import org.lenskit.specs.AbstractSpec;
+import org.lenskit.util.collections.LongUtils;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+
+/**
+ *Measure the nDPM of the top-N recommendations, using rankings.
+
+ * This metric is registered with the type name `ndpm`.
+
+ * Created by VaibhavMahant on 2/1/16.
+ */
+
+public class TopNNDPMMetric extends TopNMetric<MeanAccumulator> {
+    public static final String DEFAULT_COLUMN = "TopN.nDPM";
+    private final String columnName;
+
+    /**
+     * Construct a top-N nDCG metric from a spec.
+     * @param spec The spec.
+     */
+    @JsonCreator
+    public TopNNDPMMetric(Spec spec) {this(spec.getColumnName());
+    }
+
+
+    /**
+     * Construct a new nDPM Top-N metric.
+     * @param name The column name to use.
+     */
+    public TopNNDPMMetric(String name) {
+        super(Collections.singletonList(StringUtils.defaultString(name, DEFAULT_COLUMN)),
+              Collections.singletonList(StringUtils.defaultString(name, DEFAULT_COLUMN)));
+        columnName = StringUtils.defaultString(name, DEFAULT_COLUMN);
+    }
+    @Nullable
+    @Override
+    public MeanAccumulator createContext(AlgorithmInstance algorithm, DataSet dataSet, org.lenskit.api.Recommender recommender) {
+        return new MeanAccumulator();
+    }
+
+    @Nonnull
+    @Override
+    public MetricResult getAggregateMeasurements(MeanAccumulator context) {
+        return MetricResult.singleton(columnName, context.getMean());
+    }
+
+
+    @Nonnull
+    @Override
+    public MetricResult measureUser(TestUser user, ResultList recommendations, MeanAccumulator context) {
+        if (recommendations == null) {
+            return MetricResult.empty();
+        }
+
+        Long2DoubleMap ratings = user.getTestRatings();
+        long[] ideal = ratings.keySet().toLongArray();
+        LongArrays.quickSort(ideal, LongComparators.oppositeComparator(LongUtils.keyValueComparator(ratings)));
+
+        long[] actual = LongUtils.asLongCollection(recommendations.idList()).toLongArray();
+
+        double dpm = computeDPM(ideal, actual, ratings);
+
+        double nDPM = dpm / 2 * dpm; // Normalized nDPM
+
+        context.add(nDPM);
+        return MetricResult.singleton(columnName, nDPM);
+
+
+    }
+    /**
+     * Compute dpm of list of items, with respect to user's ratings.
+     */
+
+    double computeDPM(long [] ideal_items, long [] actual_item, Long2DoubleFunction value) {
+        double dpm;
+        int counterOne = 0;
+        int counterTwo = 0;
+        double valueOne = 0;
+        double valueTwo = 0;
+        for(int i = 0; i < actual_item.length; i++){
```

Comment:
I'm pretty sure this can be done with far less complex code. You don't need to search through `ideal_items` in order to find item IDs; you can just say `value.get(actual_item[i])`.